// This file copyright (c) 2011-2015, Quickstep Technologies LLC.
// Copyright (c) 2015, Pivotal Software, Inc.
// All rights reserved.
// See file CREDITS.txt for details.

syntax = "proto2";

package quickstep.serialization;

import "storage/StorageBlockLayout.proto";
import "types/Type.proto";
import "types/TypedValue.proto";

message CatalogAttribute {
  required string name = 1;
  required Type type = 2;
  optional string display_name = 3;
}

message PartitionScheme {
  required uint32 num_partitions = 1;
  required uint32 partition_attribute_id = 2;

  enum PartitionType {
    HASH_PARTITION = 0;
    RANGE_PARTITION = 1;
  }

  required PartitionType partition_type = 3;

  message Partition {
    repeated fixed64 blocks = 1;
  }

  repeated Partition partitions = 4;

  repeated TypedValue partition_range_boundary = 5;
}

message NUMAPlacementScheme {
  required uint32 num_numa_nodes = 1;

  message BlockToNUMANodeEntry {
    required fixed64 block_id = 1;
    required int32 numa_node = 2;
  }
  repeated BlockToNUMANodeEntry block_to_numa_node_map = 2;
}

message CatalogRelation {
  required string name = 1;
  required bool temporary = 2;

  required StorageBlockLayoutDescription default_layout = 3;

  repeated fixed64 blocks = 4 [packed=true];
  repeated CatalogAttribute attributes = 5;
  optional PartitionScheme partition_scheme = 6;
  optional NUMAPlacementScheme placement_scheme = 7;
}

message CatalogDatabase {
  required string name = 1;
  repeated CatalogRelation relations = 2;
  repeated int32 null_relations = 3;
}

message Catalog {
  repeated CatalogDatabase databases = 1;
}
